packages/code-infra/src/cli/typescript.mjs

```diff
@@ -0,0 +1,165 @@
+      } else {
+        console.error('failed to transform', dtsFile);
+      }
+    }),
+  );
+}
+
+/**
+ * Renames TypeScript declaration files.
+ * @param {Object} param0
+ * @param {string} param0.directory - The directory containing the declaration files.
+ */
+async function renameDeclarations({ directory }) {
+  const dtsFiles = await globby('**/*.d.ts', { absolute: true, cwd: directory });
+  if (dtsFiles.length === 0) {
+    return;
+  }
+  console.log(`Renaming d.ts files to d.mts in ${directory}`);
+  await Promise.all(
+    dtsFiles.map(async (dtsFile) => {
+      const newFileName = dtsFile.replace(/\.d\.ts$/, '.d.mts');
+      await fs.rename(dtsFile, newFileName);
+    }),
+  );
+}
+
+/**
+ * Creates TypeScript declaration files for the specified bundles.
+ * Types are first created in a temporary directory and then copied to the appropriate bundle directories parallelly.
+ * After copying, babel transformations are applied to the copied files because they need to be alongside the actual js files for proper resolution.
+ *
+ * @param {Object} param0
+ * @param {boolean} [param0.isMjsBuild] - Whether the build is for ESM (ECMAScript Modules).
+ * @param {{type: import('../utils/build.mjs').BundleType, dir: string}[]} param0.bundles - The bundles to create declarations for.
+ * @param {string} param0.srcDir - The source directory.
+ * @param {string} param0.buildDir - The build directory.
+ * @param {string} param0.cwd - The current working directory.
+ * @param {boolean} param0.skipTsc - Whether to skip running TypeScript compiler (tsc) for building types.
+ */
+export async function createTypes({ bundles, srcDir, buildDir, cwd, skipTsc, isMjsBuild }) {
+  const tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'code-infra-build-tsc-'));
+
+  try {
+    await copyDeclarations(srcDir, tmpDir);
+    const tsconfigPath = path.join(cwd, 'tsconfig.build.json');
+    const tsconfigExists = await fs.stat(tsconfigPath).then(
+      (file) => file.isFile(),
+      () => false,
+    );
+    if (!skipTsc) {
+      if (!tsconfigExists) {
+        throw new Error(
+          'Unable to find a tsconfig to build this project. ' +
+            `The package root needs to contain a 'tsconfig.build.json'. ` +
+            `The package root is '${cwd}'`,
+        );
+      }
+      await emitDeclarations(tsconfigPath, tmpDir);
+    }
+
+    for (const bundle of bundles) {
+      const { type: bundleType, dir: bundleOutDir } = bundle;
+      const fullOutDir = path.join(buildDir, bundleOutDir);
+      // eslint-disable-next-line no-await-in-loop
+      await fs.cp(tmpDir, fullOutDir, {
+        recursive: true,
+        force: false,
+      });
+      // eslint-disable-next-line no-await-in-loop
+      await postProcessDeclarations({
+        directory: fullOutDir,
+      });
+      if (bundleType === 'esm' && isMjsBuild) {
+        // eslint-disable-next-line no-await-in-loop
+        await renameDeclarations({
```

Comment:
Similar here, copying then renaming would probably break if cjs and esm dir are the same. Is it possible to make a copy command that can rename file son the fly, and to pass an extension (.d.ts or .d.mts to `postProcessDeclarations`)/

Perhaps I'm taking it too far?